### PR TITLE
Use offset when setting cursor position

### LINKED ISSUE #27: [Enchancement] Cursor position when dragging
### **Is your enhancement request related to a problem? Please describe.**
The current cursor position when dragging snaps to the top left of the image or text, which makes it quite difficult to position things close to the canvas boundary.

### **Describe the enhancement you'd like**
I would like the feature to maintain cursor position where it originally clicked on the image whilst dragging (similar to other tools like Canva)

### **Possible Implementation Details**
Track the offset from the image position when dragging and use that as the cursor point whilst dragging.

### **Additional context**

Current:
Screen Recording 2024-10-02 at 15 27 13

Suggested:
Screen Recording 2024-10-02 at 15 26 32

---

### **Optional Sections (if relevant):**

- **Are you willing to submit a PR for this enhancement?**   Yes
- **Does this enhancement require changes in documentation?** No

### PR DESCRIPTION
**Pull request by kvtemadden**
### **Description**
Uses offset instead of top left for cursor position whilst dragging an image.

### **Related Issue**

Closes #27 
### **Changes Made**
Used a `ref` to reference the cursor's position and used that when setting image/text position.

### **Type of Change**
_Select the type of change made with this pull request:_

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

### **Screenshots (if applicable)**
Screen Recording 2024-10-02 at 15 26 32

### **How Has This Been Tested?**
- _Manually tested on local environment_

### **Checklist**
_Check off the following items before submitting the pull request:_

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by suryanshsingh2001** [comment]
Hey @kvtemadden, it looks like your branch is a few commits behind the master branch. Please pull the latest changes to stay up to date and then add current changes.

**Comment by suryanshsingh2001** [comment]
Hey @kvtemadden,

I stumbled upon a fun little edge case—if we really push the image towards the edges of the canvas, it can straight-up vanish. Poof, gone! Not only that, but sometimes it gets a bit clingy at the borders, refusing to move smoothly. 😅

If we can smooth out these issues, it’d be amazing! Let me know if you're up for tackling this now, or if you’d prefer to open a new issue and circle back to it later.

https://github.com/user-attachments/assets/87acdc64-a28a-4f0c-97de-37158acc6a1d

**Comment by kvtemadden** [comment]
@suryanshsingh2001 - I've implemented a solution using `window` event listeners, rather than relying on the `canvas` to send events. This allows for drag outside of the canvas, which matches the behaviour of other editors (I had a go on canva to see what they do).

If someone drops the image outside of the canvas, they can use the "Reset image position" link, which sets the image back to the original position.

Thinking for a new issue it'd be cool to have an undo/redo button, but probably out of scope for this PR.

New updates:

Screen Recording 2024-10-03 at 12 32 49

File: components/shared/Editor.tsx [hunk 1]
```diff
@@ -465,6 +465,8 @@ export default function MockupEditor() {
   ]);
 
   const handleMouseDown = (e: MouseEvent) => {
+    e.preventDefault();
```
Comment (suryanshsingh2001) [inline components/shared/Editor.tsx:468]: Thank you so much for this! This bug was pretty annoying 😅

**Comment by suryanshsingh2001** [comment]
Hey @kvtemadden, everything work perfectly here, except your branch is one commit behind the `master` branch. Please take pull for latest changes and add the current changes.

**Comment by suryanshsingh2001** [comment]
> @suryanshsingh2001 - I've implemented a solution using `window` event listeners, rather than relying on the `canvas` to send events. This allows for drag outside of the canvas, which matches the behaviour of other editors (I had a go on canva to see what they do).
> 
> If someone drops the image outside of the canvas, they can use the "Reset image position" link, which sets the image back to the original position.
> 
> Thinking for a new issue it'd be cool to have an undo/redo button, but probably out of scope for this PR.
> 
> New updates:
> 
> Screen Recording 2024-10-03 at 12 32 49 [ Screen Recording 2024-10-03 at 12 32 49 ](https://private-user-images.githubusercontent.com/74623315/373235836-5620c47f-095c-432d-bab9-636f209408b8.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjc5NjIxMTMsIm5iZiI6MTcyNzk2MTgxMywicGF0aCI6Ii83NDYyMzMxNS8zNzMyMzU4MzYtNTYyMGM0N2YtMDk1Yy00MzJkLWJhYjktNjM2ZjIwOTQwOGI4LmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDAzVDEzMjMzM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTk5ZTk5MWQ0YWVhOGQxNWUyZjhiNjUwMjI0NzJjZWM4NGEwNGRiNWZmYzFjOThiOWZkMTA5OWVjNjZlYjAzNzgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.K-lOwLBWiy6MUCKHnjZvR4BMq7dw_jsYSag6jsUPNXA) [ ](https://private-user-images.githubusercontent.com/74623315/373235836-5620c47f-095c-432d-bab9-636f209408b8.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjc5NjIxMTMsIm5iZiI6MTcyNzk2MTgxMywicGF0aCI6Ii83NDYyMzMxNS8zNzMyMzU4MzYtNTYyMGM0N2YtMDk1Yy00MzJkLWJhYjktNjM2ZjIwOTQwOGI4LmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDAzVDEzMjMzM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTk5ZTk5MWQ0YWVhOGQxNWUyZjhiNjUwMjI0NzJjZWM4NGEwNGRiNWZmYzFjOThiOWZkMTA5OWVjNjZlYjAzNzgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.K-lOwLBWiy6MUCKHnjZvR4BMq7dw_jsYSag6jsUPNXA)

Yes, redo functionality is a must-have—because let's face it, we've all done something we wish we could instantly undo. Life doesn't come with a rewind button, but Mockly should! 😂

**Comment by kvtemadden** [comment]
@suryanshsingh2001, all sorted! 😇

**Comment by suryanshsingh2001** [comment]
> @suryanshsingh2001, all sorted! 😇

Hey @kvtemadden, i think all changes are not merged perfectly. I am unable to use text box anymore 🥲


https://github.com/user-attachments/assets/715f8f3a-4957-4ca6-9918-fb8099c7fc62

**Comment by kvtemadden** [comment]
Amazing spot, thank you! Just updated, text box is usable now & prevent select on drag still works :)

File: components/shared/Editor.tsx [hunk 2]
```diff
@@ -526,10 +524,12 @@ export default function MockupEditor() {
           x: x - imagePosition.x,
           y: y - imagePosition.y,
         };
+        e.preventDefault();
       } else if (text && isPointInText(x, y)) {
         setIsDragging(true);
         setDragTarget("text");
         offsetRef.current = { x: x - textPosition.x, y: y - textPosition.y };
+        e.preventDefault();
```
Comment (suryanshsingh2001) [inline components/shared/Editor.tsx:532]: What if we remove this e.preventDefault(), does the text stays inside the border but things are selectable again?

**Comment by suryanshsingh2001** [comment]
> Amazing spot, thank you! Just updated, text box is usable now & prevent select on drag still works :)

Just one small issue is happening, earlier text would stay inside canvas not it can also go out of canvas :)

**Comment by kvtemadden** [comment]
@suryanshsingh2001 what do you think would be best UX? I think text could always go outside of canvas, similar to how the image would (you'd move it bit by bit). Should I add a "reset text position" button too? Or use previous behaviour just for text?

Also re the `preventDefault`, this is what happens with and without it:

With:
Screen Recording 2024-10-03 at 20 21 40

Without:
Screen Recording 2024-10-03 at 20 24 58

As you can see, without means the text on the right gets selected whenever the mouse goes over the toolbar text.

Happy to go with whatever you think is best :)!

**Comment by suryanshsingh2001** [comment]
> @suryanshsingh2001 what do you think would be best UX? I think text could always go outside of canvas, similar to how the image would (you'd move it bit by bit). Should I add a "reset text position" button too? Or use previous behaviour just for text?
> 
> Also re the `preventDefault`, this is what happens with and without it:
> 
> With: Screen Recording 2024-10-03 at 20 21 40 [ Screen Recording 2024-10-03 at 20 21 40 ](https://private-user-images.githubusercontent.com/74623315/373388394-ae8952e5-07e7-4d33-aba8-2cbacc10514f.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjc5ODU2OTEsIm5iZiI6MTcyNzk4NTM5MSwicGF0aCI6Ii83NDYyMzMxNS8zNzMzODgzOTQtYWU4OTUyZTUtMDdlNy00ZDMzLWFiYTgtMmNiYWNjMTA1MTRmLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDAzVDE5NTYzMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBjNGJmODYxNzUzOTQ5M2U2MTIzNDFmZTRiZDhkOThjYzE3YjMzODkzMWYzZGZlMTlhNzJkMTc1ZDI0MTMzMDcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.OCkqnFbuA1-97CqEd8uLRgsRvwaSi16XtbmnneddWBo) [ ](https://private-user-images.githubusercontent.com/74623315/373388394-ae8952e5-07e7-4d33-aba8-2cbacc10514f.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjc5ODU2OTEsIm5iZiI6MTcyNzk4NTM5MSwicGF0aCI6Ii83NDYyMzMxNS8zNzMzODgzOTQtYWU4OTUyZTUtMDdlNy00ZDMzLWFiYTgtMmNiYWNjMTA1MTRmLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDAzVDE5NTYzMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBjNGJmODYxNzUzOTQ5M2U2MTIzNDFmZTRiZDhkOThjYzE3YjMzODkzMWYzZGZlMTlhNzJkMTc1ZDI0MTMzMDcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.OCkqnFbuA1-97CqEd8uLRgsRvwaSi16XtbmnneddWBo)
> 
> Without: Screen Recording 2024-10-03 at 20 24 58 [ Screen Recording 2024-10-03 at 20 24 58 ](https://private-user-images.githubusercontent.com/74623315/373388531-0126f9a5-f0de-483d-9085-a8a6ee265fe7.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjc5ODU2OTEsIm5iZiI6MTcyNzk4NTM5MSwicGF0aCI6Ii83NDYyMzMxNS8zNzMzODg1MzEtMDEyNmY5YTUtZjBkZS00ODNkLTkwODUtYThhNmVlMjY1ZmU3LmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDAzVDE5NTYzMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQ0ZDUwMzkyMmZjY2YyYzNlNGNiYjNmMWRhODJhMjA0ZDg0YzliZThhZjU0OTQxOGY0YTYwOWViYTU5OGQyNWQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.phM59H1y952eY1sbrFC6HVne4VF3Rg_zWyQ20eANFls) [ ](https://private-user-images.githubusercontent.com/74623315/373388531-0126f9a5-f0de-483d-9085-a8a6ee265fe7.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjc5ODU2OTEsIm5iZiI6MTcyNzk4NTM5MSwicGF0aCI6Ii83NDYyMzMxNS8zNzMzODg1MzEtMDEyNmY5YTUtZjBkZS00ODNkLTkwODUtYThhNmVlMjY1ZmU3LmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDAzVDE5NTYzMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQ0ZDUwMzkyMmZjY2YyYzNlNGNiYjNmMWRhODJhMjA0ZDg0YzliZThhZjU0OTQxOGY0YTYwOWViYTU5OGQyNWQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.phM59H1y952eY1sbrFC6HVne4VF3Rg_zWyQ20eANFls)
> 
> As you can see, without means the text on the right gets selected whenever the mouse goes over the toolbar text.
> 
> Happy to go with whatever you think is best :)!

You are right lets leave it as it is. But, add reset text position right next text options icon :)

**Comment by kvtemadden** [comment]
Awesome, I've already pushed that (just before your comment)

**Review by suryanshsingh2001 — APPROVED** [review]
Good job 👍

**Pull request merged**